Source/Base/ASBaseDefines.h

```diff
@@ -273,12 +269,37 @@
  * Create a new array by mapping `collection` over `work`, ignoring nil.
  */
 #define ASArrayByFlatMapping(collection, decl, work) ({ \
-  NSMutableArray *a = [NSMutableArray array]; \
+  CFTypeRef _cArray[collection.count]; \
+  NSUInteger _i = 0; \
+  for (decl in collection) {\
+    if ((_cArray[_i] = (__bridge_retained CFTypeRef)work)) { \
+      _i++; \
+    } \
+  } \
+  NSArray *result; \
+  if (_i == 0) { \
+    /** Zero array fast path. */ \
+    result = @[]; \
+  } else if (_i == 1) { \
+    /** NSSingleObjectArray is fast. Create one and release. */ \
+    CFTypeRef val = _cArray[0]; \
+    result = [NSArray arrayWithObject:(__bridge id)val]; \
+    CFBridgingRelease(val); \
+  } else { \
+    CFArrayCallBacks cb = kCFTypeArrayCallBacks; \
+    cb.retain = NULL; \
+    result = (__bridge NSArray *)CFArrayCreate(kCFAllocatorDefault, _cArray, _i, &cb); \
+  } \
```

Comment:
I'm going to factor this out a bit and clean it up, but here's the approach:
- We need to safely retain the result of the work.
- We want to transfer that retain into the array, instead of having it retain and then we release.
- The way is, use `__bridge_retained` to tell ARC "make sure this is a +1, and I'll take responsibility for it."
- Then when you create the array, use `CFArray` and provide `NULL` for the retain callback. Boom, CFArray won't retain, but it will release when the array goes away.
- Fast paths for 0- or 1- element collections e.g. `__NSArray0` and `__NSSingleObjectArray`.
- Not possible with `id []` because ARC will always release the elements when the symbol goes out of scope – there is no `CF_CONSUMED` for c-arrays.